### DISCUSSION TITLE
Concerning the move from chompjs

### DISCUSSION THREAD

**Discussion by jazzzooo** [searxng/searxng#2785] [Ideas]
So last week we switched from chompjs to pure python code in d07c006aed12631a7e0f9a5c82288a8ef7c5bfad. I do think it's an improvement to get rid of a dependency, especially when it is only used by a single engine. 

However, it already broke, so brave.news and brave.videos weren't working, which I fixed in #2784. After having a look at [the function](https://github.com/searxng/searxng/blob/master/searx/utils.py#L654) `js_variable_to_python`, I noticed it is also 5x slower than chompjs, taking around 10ms on a normal brave page, which is not bad but can definitely add up and become noticeable, especially on weaker hardware. I'm also concerned that it can come with a bigger maintenance cost than using chompjs.

Here are some other solutions:

### [chompjs](https://github.com/Nykakin/chompjs)
**Pros:**
- 400% faster than `js_variable_to_python`
- Less likely to break in the future
- Much less code on our end

**Cons:**
- External dependency written in C
- We would currently only use it for one engine

### [Function from yt-dlp](https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/utils/_utils.py#L2663)
**Pros:**
- More robust than our version
- Pure python
- Unlicensed, meaning we can use it no strings attached

**Cons:**
- Twice as slow as `js_variable_to_python`
- A lot of code in our project

### My naive implementation
I rewrote `js_variable_to_python` to be even more bare-bones, just as an example. This works exactly the same for parsing brave responses.

```python
import json
import re

QUOTE_RE = re.compile(r"([\{,])(\w+)")


def parse_js_object_naive(obj):
    parts = obj.split('"')
    for i, p in enumerate(parts[::2]):
        parts[i * 2] = QUOTE_RE.sub(r'\1"\2"', p.replace(":.", ":0.").replace("void 0", "null"))
    return json.loads('"'.join(parts))
```
**Pros:**
- 50% faster than `js_variable_to_python`
- 6 lines of code instead of 70
- Potentially less maintenance cost
- Potentially easier to understand

**Cons:**
- Makes more assumptions and is just as likely to break as `js_variable_to_python`
- Pretty hacky

## Testing
```
In [1]: import chompjs

In [2]: %timeit -r 15 js_variable_to_python(data)
9.31 ms ± 7.75 µs per loop (mean ± std. dev. of 15 runs, 100 loops each)

In [3]: %timeit -r 15 parse_js_object_naive(data)
6.15 ms ± 53.6 µs per loop (mean ± std. dev. of 15 runs, 100 loops each)

In [4]: %timeit -r 15 chompjs.parse_js_object(data)
1.82 ms ± 2.25 µs per loop (mean ± std. dev. of 15 runs, 100 loops each)

In [5]: js_variable_to_python(data) == parse_js_object_naive(data)
Out[5]: True

```

**Comment by return42** [comment]
Further solutions & discussions: https://stackoverflow.com/questions/24027589/how-to-convert-raw-javascript-object-to-a-dictionary

**Reply by jazzzooo** [reply]
Nice! Demjson is a dead project, ast.literal_eval() or simply calling eval() both fail with `{a: .5}`. I'll try jsonnet later, maybe its faster than chompjs

**Comment by dalf** [comment]
Since I'm of the author of https://github.com/searxng/searxng/commit/d07c006aed12631a7e0f9a5c82288a8ef7c5bfad here some context.

An additional cons about chompjs: it is maintained by one person and parse external data using C. The [tests](https://github.com/Nykakin/chompjs/blob/master/chompjs/test_parser.py) might be enough or it, the code enough simple, but I'm afraid of a buffer overflow. curl, lxml are also written in C, but there are far more contributors and eyes on the code.

I wrote js_variable_to_python for the google image engine one or two years ago, but actually the engine never used it.

As far I understand, parse_js_object_naive does not parse escape sequence. It will break as soon a title or content contains double quotes.

The function from yt-dlp has a `strict=False` parameter. `strict=True` removes some regex which might speed up the calls. [This gist](https://gist.github.com/dalf/dc9eccae428089bc636a8a052622aad5#file-js_to_json-py-L1-L76) contains a version without template support (it is just a quick prototype).

I'm currently adding the tests from chompjs to check the function js_variable_to_python to investigate. I'm not saying we should continue to use this function, my stance is to keep this function has long only one engine need it and it does not break too often which might happen, yes as you are saying. So, it is good thing to look for alternatives.

**Reply by return42** [reply]
> So, it is good thing to look for alternatives.

:+1: .. I think its good to look out for a robust implementation since I expect the need of such function in future implementation (engines) .. and then it is good to have a solution which works well with a wide variation of JS JSON-strings .. further I agree with @dalf that this solution should not be based on chompjs (for the reasons he named) .. thanks to you all for having an eye on this.

**Reply by jazzzooo** [reply]
Thank you for your input!

> As far I understand, parse_js_object_naive does not parse escape sequence. 

Yep, it makes some huge assumptions, but it serves a lower bound for performance in pure python. Unless someone can make it even faster :smile:

Also if we keep using js_variable_to_python (and maybe parse_js_object would be a more appropriate name for it), I also have some safe and some less safe ways to increase it's performance 5-20%.

@return42 do you have any opinion on how fast searxng needs to be? I think 100ms for parsing would be too long, but maybe 10ms is acceptable? [Some claim](https://web.dev/rail/#50-ms-or-100-ms) 50ms should be the [standard](https://developer.mozilla.org/en-US/docs/Web/Performance/How_long_is_too_long).

**Reply by return42** [reply]
> do you have any opinion on how fast searxng needs to be? 

No, I think we have to find a compromise in between robustness, security reasons (not C) and performance.
